File: eng/docs/agent_skills/csharp_typing/examples/integration_tests/FailurePropagationTest.cs
// ============================================================
// Integration test: Failure propagation
// Family: failure_propagation
// Level: cmp_int (component integration — multi-step flow)
// ============================================================
//
// Verifies that an early failure in one step prevents
// downstream steps from executing and produces the correct
// result and observability events.

using FluentAssertions;
using Xunit;

// Adjust these usings to your real namespace.

namespace Acme.ExecutionEngine.Tests.Integration;

// ============================================================
// Domain types (self-contained example)
// ============================================================

internal sealed record ExecutionOutcome(string Status, string Reason);

// ============================================================
// Event capture
// ============================================================

internal sealed class FailurePropagationEventCapture
{
    private readonly List<Dictionary<string, object>> _events = [];

    public void Emit(Dictionary<string, object> ev) => _events.Add(ev);

    public List<string> EventNames =>
        _events.Select(e => (string)e["ev"]).ToList();
}

// ============================================================
// Collaborators
// ============================================================

internal sealed class FailingAgentStore
{
    public Dictionary<string, object> ResolveAgent(string agentId)
    {
        _ = agentId;
        throw new KeyNotFoundException("AGENT_RESOLUTION_FAILED");
    }
}

internal sealed class GenerationSpy
{
    public int Calls { get; private set; }

    public string Generate()
    {
        Calls++;
        return "generated";
    }
}

// ============================================================
// Engine under test
// ============================================================

internal sealed class ExecutionEngine
{
    private readonly FailingAgentStore _agentStore;
    private readonly GenerationSpy _generator;
    private readonly FailurePropagationEventCapture _obs;

    public ExecutionEngine(
        FailingAgentStore agentStore,
        GenerationSpy generator,
        FailurePropagationEventCapture obs)
    {
        _agentStore = agentStore;
        _generator = generator;
        _obs = obs;
    }

    public ExecutionOutcome Execute(string agentId)
    {
        _obs.Emit(new() { { "ev", "EXECUTION_STARTED" } });
        try
        {
            _agentStore.ResolveAgent(agentId);
        }
        catch (KeyNotFoundException)
        {
            _obs.Emit(new() { { "ev", "AGENT_RESOLVE_FAILED" }, { "severity", "WARN" } });
            return new ExecutionOutcome(
                Status: "RETRYABLE_FAIL",
                Reason: "AGENT_RESOLUTION_FAILED");
        }

        _generator.Generate();
        _obs.Emit(new() { { "ev", "EXECUTION_DONE" } });
        return new ExecutionOutcome(Status: "DONE", Reason: "SUCCESS");
    }
}

// ============================================================
// Tests
// ============================================================

[Trait("Level", "cmp_int")]
[Trait("Family", "failure_propagation")]
public sealed class FailurePropagationTest
{
    [Fact]
    public void ReturnsRetryableFailAndSkipsGenerationWhenAgentResolutionFails()
    {
        // Arrange
        var obs = new FailurePropagationEventCapture();
        var gen = new GenerationSpy();
        var engine = new ExecutionEngine(
            new FailingAgentStore(), gen, obs);

        // Act
        var result = engine.Execute("agent_a");

        // Assert — correct result
        result.Should().Be(
            new ExecutionOutcome(
                Status: "RETRYABLE_FAIL",
                Reason: "AGENT_RESOLUTION_FAILED"));

        // Assert — generation was skipped
        gen.Calls.Should().Be(0);

        // Assert — observability events in correct order
        obs.EventNames.Should().BeEquivalentTo(
            new[] { "EXECUTION_STARTED", "AGENT_RESOLVE_FAILED" },
            options => options.WithStrictOrdering());
    }
}

File: eng/docs/agent_skills/python_typing/examples/integration tests/failure_propagation_test.py
from __future__ import annotations

from dataclasses import dataclass

import pytest

from dictionary.implementation.TL_CORE import Result


@dataclass(frozen=True, slots=True)
class ExecutionOutcome:
    status: str
    reason: str


class EventCapture:
    def __init__(self) -> None:
        self.events: list[dict] = []

    def emit(self, event: dict) -> None:
        self.events.append(event)


class FailingAgentStore:
    def resolve_agent(self, agent_id: str) -> dict:
        del agent_id
        raise LookupError("AGENT_RESOLUTION_FAILED")


class GenerationSpy:
    def __init__(self) -> None:
        self.calls = 0

    def generate(self) -> str:
        self.calls += 1
        return "generated"


class ExecutionEngine:
    def __init__(self, agent_store: FailingAgentStore, generator: GenerationSpy, obs: EventCapture) -> None:
        self._agent_store = agent_store
        self._generator = generator
        self._obs = obs

    def execute(self, agent_id: str) -> Result[ExecutionOutcome]:
        self._obs.emit({"ev": "EXECUTION_STARTED"})
        try:
            self._agent_store.resolve_agent(agent_id)
        except LookupError:
            self._obs.emit({"ev": "AGENT_RESOLVE_FAILED", "severity": "WARN"})
            return Result(
                outcome="ok",
                ok=ExecutionOutcome(
                    status="RETRYABLE_FAIL",
                    reason="AGENT_RESOLUTION_FAILED",
                ),
            )

        self._generator.generate()
        self._obs.emit({"ev": "EXECUTION_DONE"})
        return Result(
            outcome="ok",
            ok=ExecutionOutcome(status="DONE", reason="SUCCESS"),
        )


@pytest.fixture
def failure_slice() -> tuple[ExecutionEngine, GenerationSpy, EventCapture]:
    obs = EventCapture()
    gen = GenerationSpy()
    ee = ExecutionEngine(agent_store=FailingAgentStore(), generator=gen, obs=obs)
    return ee, gen, obs


@pytest.mark.cmp_int
@pytest.mark.failure_propagation
def test_execution_returns_retryable_fail_and_skips_generation_when_agent_resolution_fails(
    failure_slice: tuple[ExecutionEngine, GenerationSpy, EventCapture],
) -> None:
    ee, gen, obs = failure_slice

    result = ee.execute("agent_a")

    assert result.outcome == "ok"
    assert result.ok == ExecutionOutcome(
        status="RETRYABLE_FAIL",
        reason="AGENT_RESOLUTION_FAILED",
    )
    assert gen.calls == 0
    assert [e["ev"] for e in obs.events] == [
        "EXECUTION_STARTED",
        "AGENT_RESOLVE_FAILED",
    ]
